File: modules/item-classifier/ios/PerceptualHash.swift
import Accelerate
import CoreGraphics
import UIKit

/// Perceptual hashing using DCT for image similarity matching.
enum PerceptualHash {
  private static let hashSize = 32 // resize target
  private static let lowFreq = 8   // 8×8 low-frequency block

  /// Compute a 64-bit perceptual hash from a CGImage.
  static func compute(from cgImage: CGImage) -> UInt64 {
    // Convert to 32×32 grayscale
    let size = hashSize
    let grayscale = toGrayscale(cgImage, size: size)

    // 2D DCT via Accelerate
    let dctResult = dct2d(grayscale, size: size)

    // Extract 8×8 low-frequency block (skip DC at [0,0])
    var lowFreqValues: [Float] = []
    lowFreqValues.reserveCapacity(lowFreq * lowFreq - 1)
    for r in 0..<lowFreq {
      for c in 0..<lowFreq {
        if r == 0 && c == 0 { continue }
        lowFreqValues.append(dctResult[r * size + c])
      }
    }

    // Median threshold
    let sorted = lowFreqValues.sorted()
    let median = sorted[sorted.count / 2]

    // Build 64-bit hash
    var hash: UInt64 = 0
    for i in 0..<min(64, lowFreqValues.count) {
      if lowFreqValues[i] > median {
        hash |= (1 << i)
      }
    }

    return hash
  }

  /// Hamming distance between two hashes (number of differing bits).
  static func hammingDistance(_ a: UInt64, _ b: UInt64) -> Int {
    return (a ^ b).nonzeroBitCount
  }

  // MARK: - Private

  private static func toGrayscale(_ image: CGImage, size: Int) -> [Float] {
    let colorSpace = CGColorSpaceCreateDeviceGray()
    var pixels = [UInt8](repeating: 0, count: size * size)

    guard let context = CGContext(
      data: &pixels,
      width: size,
      height: size,
      bitsPerComponent: 8,
      bytesPerRow: size,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.none.rawValue
    ) else {
      return [Float](repeating: 0, count: size * size)
    }

    context.draw(image, in: CGRect(x: 0, y: 0, width: size, height: size))

    return pixels.map { Float($0) }
  }

  private static func dct2d(_ input: [Float], size: Int) -> [Float] {
    var result = [Float](repeating: 0, count: size * size)

    // DCT on rows
    var rowDCT = [Float](repeating: 0, count: size * size)
    for r in 0..<size {
      var row = Array(input[(r * size)..<((r + 1) * size)])
      var output = [Float](repeating: 0, count: size)

      let setup = vDSP_DCT_CreateSetup(nil, vDSP_Length(size), .II)!
      vDSP_DCT_Execute(setup, &row, &output)
      vDSP_DFT_DestroySetup(setup)

      for c in 0..<size {
        rowDCT[r * size + c] = output[c]
      }
    }

    // DCT on columns
    for c in 0..<size {
      var col = [Float](repeating: 0, count: size)
      for r in 0..<size {
        col[r] = rowDCT[r * size + c]
      }
      var output = [Float](repeating: 0, count: size)

      let setup = vDSP_DCT_CreateSetup(nil, vDSP_Length(size), .II)!
      vDSP_DCT_Execute(setup, &col, &output)
      vDSP_DFT_DestroySetup(setup)

      for r in 0..<size {
        result[r * size + c] = output[r]
      }
    }

    return result
  }
}

File: modules/item-classifier/src/android/perceptual-hash.ts
/**
 * Perceptual hash (pHash) computation for Android.
 *
 * Ported from scripts/build-fingerprints.ts — same DCT-based 64-bit hash
 * algorithm so results are compatible with the pre-computed fingerprints.
 */
import type { PixelData } from "./image-utils";

const PHASH_SIZE = 32;
const DCT_LOW = 8;

/** Type-II DCT on an array of length N (naive O(N^2), fine for N=32). */
function dct1d(input: Float64Array): Float64Array {
  const N = input.length;
  const output = new Float64Array(N);
  for (let k = 0; k < N; k++) {
    let sum = 0;
    for (let n = 0; n < N; n++) {
      sum += input[n] * Math.cos((Math.PI * (2 * n + 1) * k) / (2 * N));
    }
    output[k] = sum;
  }
  return output;
}

/** 2D DCT on a row-major matrix. */
function dct2d(matrix: Float64Array, size: number): Float64Array {
  const result = new Float64Array(size * size);

  // DCT on rows
  const rowDCT = new Float64Array(size * size);
  const rowBuf = new Float64Array(size);
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) rowBuf[c] = matrix[r * size + c];
    const transformed = dct1d(rowBuf);
    for (let c = 0; c < size; c++) rowDCT[r * size + c] = transformed[c];
  }

  // DCT on columns
  const colBuf = new Float64Array(size);
  for (let c = 0; c < size; c++) {
    for (let r = 0; r < size; r++) colBuf[r] = rowDCT[r * size + c];
    const transformed = dct1d(colBuf);
    for (let r = 0; r < size; r++) result[r * size + c] = transformed[r];
  }

  return result;
}

/**
 * Compute a 64-bit perceptual hash from RGBA pixel data.
 *
 * The input should be 32x32 pixels (already resized).
 * Returns the hash as two 32-bit numbers [hi, lo] since JS doesn't
 * have native uint64 support in React Native (no BigInt in Hermes).
 */
export function computePHash(pixels: PixelData): [number, number] {
  const { width, height, data } = pixels;

  // Convert to grayscale float64
  const gray = new Float64Array(width * height);
  for (let i = 0; i < width * height; i++) {
    const r = data[i * 4];
    const g = data[i * 4 + 1];
    const b = data[i * 4 + 2];
    gray[i] = 0.299 * r + 0.587 * g + 0.114 * b;
  }

  // 2D DCT
  const dctResult = dct2d(gray, PHASH_SIZE);

  // Extract top-left 8x8 low-frequency coefficients (skip DC at [0,0])
  const lowFreq: number[] = [];
  for (let r = 0; r < DCT_LOW; r++) {
    for (let c = 0; c < DCT_LOW; c++) {
      if (r === 0 && c === 0) continue;
      lowFreq.push(dctResult[r * PHASH_SIZE + c]);
    }
  }

  // Median threshold
  const sorted = [...lowFreq].sort((a, b) => a - b);
  const median = sorted[Math.floor(sorted.length / 2)];

  // Build 64-bit hash as [hi32, lo32]
  let lo = 0;
  let hi = 0;
  for (let i = 0; i < 64; i++) {
    if (i < lowFreq.length && lowFreq[i] > median) {
      if (i < 32) {
        lo |= 1 << i;
      } else {
        hi |= 1 << (i - 32);
      }
    }
  }

  return [hi, lo];
}

/**
 * Hamming distance between two 64-bit hashes (represented as [hi, lo] pairs).
 * Returns number of differing bits (0 = identical, 64 = maximally different).
 */
export function hammingDistance(
  a: [number, number],
  b: [number, number]
): number {
  return popcount(a[0] ^ b[0]) + popcount(a[1] ^ b[1]);
}

/** Count set bits in a 32-bit integer. */
function popcount(n: number): number {
  n = n - ((n >> 1) & 0x55555555);
  n = (n & 0x33333333) + ((n >> 2) & 0x33333333);
  return (((n + (n >> 4)) & 0x0f0f0f0f) * 0x01010101) >> 24;
}

/**
 * Convert a uint64 LE from a DataView into [hi, lo] pair.
 * Used when parsing sprite-fingerprints.bin.
 */
export function readUint64LE(
  view: DataView,
  offset: number
): [number, number] {
  const lo = view.getUint32(offset, true);
  const hi = view.getUint32(offset + 4, true);
  return [hi, lo];
}
